Port to .NET Core 3

0xced opened the pull request:
Here's a pull request porting ICanHasDotnetCore to .NET Core 3 with a few side improvements. I have tried to break things in isolated commits so in case you don't want to merge it altogether you can cherry-pick commits of interest.

Here are the most important changes:
* Use Entity Framework Core instead of raw SQL
This change makes it super easy to develop with SQLite instead of SQL Server (I did all the work on macOS using JetBrains Rider). To develop on SQLite, just add this in `appsettings.Development.json`:
```json
{
  "Database": {
    "Provider": "Sqlite",
    "ConnectionString": "DataSource=/path/to/ICanHasDotnetCore.sqlite3"
  }
}
```

And maybe you could even use SQLite in production and save money on an SQL Server instance?

* Use [NuGet.Protocol](https://www.nuget.org/packages/NuGet.Protocol/) instead of NuGet.Core which is not compatible with .NET Core. This change has brought to light some packages that were incorrectly identified, the tests have been adapted accordingly.
* Improved logs overall. The logs added for Octokit allowed me to identify how to [improve performance of GitHub scanning](https://github.com/0xced/ICanHasDotnetCore/commit/689bb3d4b71de88ff263b1f89b8bef4f4423f55e).
* Web project modernization

Updated the web project to work with Node.js 12 and Gulp 4. I also added a new *Scan A NuGet package* tab to get the dependency graph of a NuGet package without having to upload a file.

And here are less important changes:

* Updated NuGet packages to their latest versions
* Replaced AutoFac with built-in ASP.NET Core equivalent features
* Replace Newtonsoft.Json with System.Text.Json
* Asyncified all code
* Updated knowledge
* Fixed a potential `ArgumentNullException`

0xced committed:
Setup UseDeveloperExceptionPage as the first middleware

Having app.UseDeveloperExceptionPage as the last middleware won't catch any exception.

See https://docs.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-2.2#the-developer-exception-page

0xced committed:
Update System.Xml.* to latest version (4.3.0)

0xced committed:
Update dbup to version 4.2.0

0xced committed:
Add strongly-typed configuration classes

0xced committed:
Fix solution items paths

0xced committed:
Update Octokit to version 0.36.0

This fixes GitHubScannerTests.ICanHasDotnetRepository and GitHubScannerTests.RepoDoesNotExist

0xced committed:
Update Autofac.Extensions.DependencyInjection to version 5.0.1

0xced committed:
Update to version 2.2.0 of ASP.NET Core

0xced committed:
Update Newtonsoft.JSON to version 12.0.2

0xced committed:
Update Serilog to latest versions

Also replace the retired Serilog.Sinks.ColoredConsole and Serilog.Sinks.LiterateConsole with Serilog.Sinks.Console

0xced committed:
Update tests NuGet packages to latest versions

0xced committed:
Use Microsoft provided https redirection instead of custom one

Also use the IMiddleware interface for RedirectWwwMiddleware

0xced committed:
Remove the default console debug log and use Serilog only instead

Logs were appearing twice in the console.

0xced committed:
Move configuration into the web host builder

See also discussion on https://github.com/serilog/serilog-settings-configuration/issues/66#issuecomment-325419790

0xced committed:
Add links to Edotor and Graphviz Online

The nice thing with Edotor and Graphviz Online is that you can pass the whole graph in the URL.

0xced committed:
Fix Graphviz spelling

0xced committed:
Modernize web project to work with node 12

* Use @types packages instead of typings
* Add more typescript types

0xced committed:
Convert to Entity Framework Core

0xced committed:
Upgrade to Entity Framework Core 3.0.0

0xced committed:
Implement Requery as BackgroundService instead of Autofac.IStartable

0xced committed:
Add "Scan A NuGet package" tab

Useful for generating a dependency graph of a single existing NuGet package

0xced committed:
Upgrade to ASP.NET Core 3.0

0xced committed:
Improve parallelization of tests

This way the xunit test runner is not confused. Idea for parallelization from 
https://stackoverflow.com/questions/32702969/how-to-make-xunit-to-run-a-theory-parallel/34828001#34828001

0xced committed:
Asyncify everything that should be async

0xced committed:
Replace Autofac with the builtin DI container

0xced committed:
Replace Newtonsoft.Json with System.Text.Json

0xced committed:
Update knowledge

* NSubstitute is now compatible with .NET Core
* HtmlAgilityPack is now compatible with .NET Core
* iTextSharp has been officially replaced by iText 7

0xced committed:
Improve logs

0xced committed:
Improve GitHub scanning performance

By specifying PageCount = 1 and PageSize = 1 we fetch only the last commit, potentially sparing a lot of HTTP requests to the GitHub API when the repository has a long history.

0xced committed:
Fix ArgumentNullException

Stemming from `PackageReference` with `Remove` attribute instead of `Include` attribute in csproj files.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/OctopusDeploy/ICanHasDotnetCore?pullRequest=40) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/OctopusDeploy/ICanHasDotnetCore?pullRequest=40) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OctopusDeploy/ICanHasDotnetCore?pullRequest=40) it.</sub>

0xced committed:
More support for cancellation token

* NugetPackageInfoRetriever
* PackageRepositoryWrapper

0xced committed:
Add known replacement for NuGet.Core

0xced committed:
Add possibility to give an explicit git reference when scanning GitHub

Scanning a GitHub repository now supports the following syntax: Username/Repository@Reference with the last part "@Reference" being optional. This allows to scan a repository for a given tag, branch or specific commit with its SHA1.

0xced committed:
Always return the result of the pre-release package if we search one

This solves the case of "Not a .NET Library" being only available in pre-release such as Microsoft.NETFramework.ReferenceAssemblies.

Before this commit it would diagnose Microsoft.NETFramework.ReferenceAssemblies as "Not Found". After this commit, it is correctly diagnosed as "Not a .NET Library".

0xced committed:
Use AsNoTracking() for read-only scenarios

0xced committed:
Add RootNamespace to all projects

This fixes ReSharper/Rider warnings

0xced committed:
Refactor MoreInformationRepository and KnownReplacementsRepository

They now share a common implementation (EmbeddedResourceRepository) and are not static anymore

0xced committed:
Reduce log noise when cancelling a request

0xced committed:
Use NuGet.Protocol instead of NuGet.Core

This is a necessary step to convert the project to .NET Core

0xced committed:
Convert to using declarations

And use await using for IAsyncDisposable

droyad commented on source/Web/Features/Statistics/StatisticsRepository.cs:
```diff
@@ -1,125 +1,79 @@
-
-            try
+            await using var context = _contextFactory();
+            var packageStatistic = await context.PackageStatistics.FindAsync(new object[] {package.PackageName}, cancellationToken);
```
This is no longer atomic. I used the `MERGE` command as it guaranteed (from my understanding) atomic insert/update of that record. With this change two threads could update the same statistic record and only the last in wins.

0xced replied:
Good catch! I addressed this issue in 49561dba1e6f68497d69efe13669c1c8e0e0b6d1. I used [FlexLabs.EntityFrameworkCore.Upsert](https://github.com/artiomchi/FlexLabs.Upsert) as EF Core does not handle upsert out of the box.

By increasing the EF Core log level to information (`.MinimumLevel.Override("Microsoft.EntityFrameworkCore", LogEventLevel.Information)`) and enabling sensitive data logging (`optionsBuilder.EnableSensitiveDataLogging();`) we can see how FlexLabs.Upsert works:

With SQLite:

```sql
-- Executing DbCommand [Parameters=[@p0='Newtonsoft.Json' (Nullable = false) (Size = 15), @p1='1' (DbType = String), @p2='Supported' (Nullable = false) (Size = 9), @p3='1' (DbType = String)], CommandType='Text', CommandTimeout='30']
INSERT INTO "PackageStatistics" AS "T" ("Name", "Count", "LatestSupportType") VALUES (@p0, @p1, @p2) ON CONFLICT ("Name") DO UPDATE SET "Count" = ( "T"."Count" + @p3 )
```

With SQL Server:

```sql
-- Executing DbCommand [Parameters=[@p0='Newtonsoft.Json' (Nullable = false) (Size = 450), @p1='1', @p2='Supported' (Nullable = false) (Size = 4000), @p3='1'], CommandType='Text', CommandTimeout='30']
MERGE INTO [PackageStatistics] WITH (HOLDLOCK) AS [T] USING ( VALUES (@p0, @p1, @p2) ) AS [S] ([Name], [Count], [LatestSupportType]) ON [T].[Name] = [S].[Name] WHEN NOT MATCHED BY TARGET THEN INSERT ([Name], [Count], [LatestSupportType]) VALUES ([Name], [Count], [LatestSupportType]) WHEN MATCHED THEN UPDATE SET [Count] = ( [T].[Count] + @p3 );
```

droyad commented:
@0xced I haven't forgotten about this and will get to doing a full review. I did a quick scan and only one thing stood out to me so far (added a comment)

0xced committed:
Make package statistic upsert atomic

EF Core does not support upsert out of the box, see https://github.com/aspnet/EntityFrameworkCore/issues/4526

But Artiom Chilaru created a NuGet package (FlexLabs.EntityFrameworkCore.Upsert) that handle upsert for PostgreSQL/Sqlite, SqlServer and MySQL.

0xced commented:
Excellent, I’ll stop force-pushing small changes while you are reviewing.

Would you consider using SQLite in production? I think for a simple two tables database (cache + stats) it would be well enough and cheaper than an SQL Server instance. Are you currently using an Azure SQL Database? Is it costly?

0xced committed:
Ensure the directory is created when using an SQLite database

0xced committed:
Upgrade from netcoreapp3.0 to netcoreapp3.1

0xced committed:
Use the default Octokit http message handler

The default handler properly handle authenticated redirects

0xced committed:
Log rate limit information for Octokit

0xced committed:
Add an HTTP cache to Octokit

This improves performance and also helps staying within the GitHub rate limit: https://developer.github.com/v3/#staying-within-the-rate-limit

0xced committed:
Code cleanup: fix double spacing

0xced committed:
Code cleanup: remove unused directives

0xced committed:
Update all NuGet packages to latest versions

0xced committed:
Fix new EF Core warning

See https://github.com/dotnet/efcore/issues/17471#issuecomment-526330450

0xced committed:
Code cleanup: fix all ReSharper warnings

0xced committed:
Improve performance by fetching raw data from GitHub

0xced commented:
I have pushed a few more improvements. We could probably shave a few more milliseconds when scanning GitHub repositories if https://github.com/octokit/octokit.net/pull/2151 gets merged and a new release of Octokit is released.

0xced committed:
Use Ok() instead of Json() for the GitHub action

Also make the return type explicit with Task<ActionResult<GetResultResponse>>

0xced committed:
Add statistics when computing dependencies on a GitHub repository

0xced committed:
Upgrade NuGet packages to their latest versions

0xced committed:
Simplify creation of host and web host

droyad commented:
@0xced Thank you for this contribution. Unfortunately we've decided to shutter the site. The traffic has decreased significantly since .NET Core 3.1 was released (and previous 2.0). The ecosystem as a whole has transitioned over and the majority of use cases now work.
